README.md
# react-button-a11y
[![All Contributors](https://img.shields.io/badge/all_contributors-1-orange.svg?style=flat-square)](#contributors)
[![npm version](https://img.shields.io/npm/v/react-button-a11y.svg)](https://www.npmjs.com/package/react-button-a11y) [![npm downloads](https://img.shields.io/npm/dm/react-button-a11y.svg)](https://www.npmjs.com/package/react-button-a11y) [![Build Status](https://travis-ci.org/rpearce/react-button-a11y.svg?branch=master)](https://travis-ci.org/rpearce/react-button-a11y) [![Coverage Status](https://coveralls.io/repos/github/rpearce/react-button-a11y/badge.svg?branch=master)](https://coveralls.io/github/rpearce/react-button-a11y?branch=master) [![Maintainability](https://api.codeclimate.com/v1/badges/8e4debef4b9f0e8acd6e/maintainability)](https://codeclimate.com/github/rpearce/react-button-a11y/maintainability)

Make non-button elements accessible in React. This component applies the
following to an element that wants to be a "button":
  * `role="button"`
  * `tabIndex="0"`
  * `onKeyDown` listener for enter and spacebar (like normal `<button>`s) that
    will trigger an `onClick` callback
  * enforced labelling via either `aria-label` or `children`

## Links
* [Installation](#installation)
* [Usage](#usage)
* [API](#api)
* [All Contributors](#contributors)
* [Authors](./AUTHORS)
* [Changelog](./CHANGELOG.md)
* [Contributing](./CONTRIBUTING.md)
* [Code of Conduct](./CODE_OF_CONDUCT.md)

# Installation
```
npm i react-button-a11y
```
or
```
yarn add react-button-a11y
```

## Usage

### With `aria-label`
```js
import ButtonA11y from 'react-button-a11y'

// ...

<ButtonA11y
  aria-label="Click this to do X"
  className="some-className"
  onClick={evt => { console.log('press happened: ', evt) }}
/>
```

### Without `aria-label`
If you don't provide an `aria-label`, you need to provide `children` so that
there is something to describe what clicking / tapping this button does.
```js
import ButtonA11y from 'react-button-a11y'

// ...

<ButtonA11y
  className="some-className"
  onClick={evt => { console.log('press happened: ', evt) }}
>
  Click this to do X
</ButtonA11y>
```

# API

| Prop | Type | Required | Default  | Details |
| ---  | --- | ---  | --- | --- |
| `aria-label` | String | yes if `children` absent | none | Pass an `aria-label` if you don't pass children with descriptive text |
| `children` | Node | yes if `aria-label` absent | none | Pass `children` with descriptive text if you don't pass an `aria-label` |
| `element` | String | no | `'span'` | The default element is a `div`, but you can pass a `span`, e.g., if you like |
| `onClick` | Function | yes | `Function.prototype` | When the button is "pressed" (via Enter or Spacebar or click), this callback will be triggered with the event |
| `strictMode` | Bool | no | `true` | An error will be thrown if enabled and you fail to pass both an `aria-label` and `children` |

## Contributors

Thanks goes to these wonderful people ([emoji key](https://github.com/kentcdodds/all-contributors#emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
| [<img src="https://avatars2.githubusercontent.com/u/592876?v=4" width="100px;"/><br /><sub><b>Robert Pearce</b></sub>](https://robertwpearce.com)<br />[💻](https://github.com/rpearce/react-button-a11y/commits?author=rpearce "Code") [📖](https://github.com/rpearce/react-button-a11y/commits?author=rpearce "Documentation") [💡](#example-rpearce "Examples") [🤔](#ideas-rpearce "Ideas, Planning, & Feedback") [⚠️](https://github.com/rpearce/react-button-a11y/commits?author=rpearce "Tests") |
| :---: |
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/kentcdodds/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [4.1.2] - 2020-01-01

### Changed
* fixed commonjs bundle

## [4.1.1] - 2019-12-29

### Changed
* bump `react-with-forwarded-ref`

## [4.1.0] - 2019-12-29

### Added
* added `browser` field for umd build

### Changed
* bumped all deps
* updated `main` and `module` fields
* license from ISC to BSD-3
* moved API docs back to README

### Removed
* `.browserslistrc` file (just use babel preset-env default)

## [4.0.0] - 2019-09-12

### Changed
* now building commonjs & ESM build outputs in `dist/` folder
* using rollup for building output
* security fixes
* removed peer dependency on `react-dom`

## [3.0.1] - 2019-02-13

### Added
* added missing .browserslistrc file

### Changed
* bumped some dev dependencies

## [3.0.0] - 2019-02-13

### Changed
* sets a new peer dependency to be at least React 16.8
* converts component to functional component and uses the [`useCallback`
  hook](https://reactjs.org/docs/hooks-reference.html#usecallback) for the
  `handleKeyDown` event memoization
* wraps the component and forwarded ref higher-order component in
  [`React.memo`](https://reactjs.org/docs/react-api.html#reactmemo)
* updates example to use [`useState`](https://reactjs.org/docs/hooks-reference.html#usestate)
  and [`useCallback`](https://reactjs.org/docs/hooks-reference.html#usecallback)
  to toggle the button states (text and background color, where applicable)
* rewrites the tests in
  [`react-testing-library`](https://github.com/kentcdodds/react-testing-library)
  to focus less on internals and more on what an end-user might see (and really
  to try it out, too)

## [2.0.0] - 2018-12-29

### Changed
* changed default `element` from `div` to `span`. Unfortunately, this qualifies
  as a breaking change (hence the major version bump).

## [1.0.1] - 2018-12-19

### Fixed
* fixed forwarding refs eating the display name by including
  `react-with-forwarded-refs` as a dependency

## [1.0.0] - 2018-12-18

### Fixed
* support for forwarding refs (#5)

### Changed
* peer dependency on react has to be bumped to `>= 16.3`, so that breaking
  change is the reason for the major version bump

## [0.1.0] - 2018-12-14

### Added
* Added all the things


__tests__/index.js
import React, { createRef } from 'react'
import { fireEvent, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import ButtonA11y from '../source'

test('renders with aria-label, handles click', () => {
  const onClick = jest.fn()
  const { getByLabelText } = render(
    <ButtonA11y
      aria-label="Mock Label"
      className="mock-className"
      onClick={onClick}
    />
  )
  const el = getByLabelText('Mock Label')

  expect(el).toBeInTheDocument()

  fireEvent.click(el)

  expect(onClick).toHaveBeenCalledWith(expect.any(Object))
})

test('renders with children, handles click', () => {
  const onClick = jest.fn()
  const { getByText } = render(
    <ButtonA11y className="mock-className" onClick={onClick}>
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.click(el)

  expect(onClick).toHaveBeenCalledWith(expect.any(Object))
})

test('renders with button role and tabIndex 0', () => {
  const onClick = jest.fn()
  const { getByText } = render(
    <ButtonA11y className="mock-className" onClick={onClick}>
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()
  expect(el.getAttribute('role')).toEqual('button')
  expect(el.getAttribute('tabindex')).toEqual('0')
})

test('calls onKeyDown and onClick when key is Enter', () => {
  const onClick = jest.fn()
  const onKeyDown = jest.fn()
  const { getByText } = render(
    <ButtonA11y
      className="mock-className"
      onKeyDown={onKeyDown}
      onClick={onClick}
      type=""
    >
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { key: 'Enter' })

  expect(onKeyDown).toHaveBeenCalledWith(expect.any(Object))
  expect(onClick).toHaveBeenCalledWith(expect.any(Object))
})

test('calls onKeyDown and onClick when key is empty space', () => {
  const onClick = jest.fn()
  const onKeyDown = jest.fn()
  const { getByText } = render(
    <ButtonA11y
      className="mock-className"
      onKeyDown={onKeyDown}
      onClick={onClick}
      type=""
    >
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { key: ' ' })

  expect(onKeyDown).toHaveBeenCalledWith(expect.any(Object))
  expect(onClick).toHaveBeenCalledWith(expect.any(Object))
})

test('calls onKeyDown and onClick when key is Spacebar', () => {
  const onClick = jest.fn()
  const onKeyDown = jest.fn()
  const onScroll = jest.fn()
  const { getByText } = render(
    <ButtonA11y
      className="mock-className"
      onKeyDown={onKeyDown}
      onClick={onClick}
      type=""
    >
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  global.document.addEventListener('scroll', onScroll)

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { key: 'Spacebar' })

  expect(onKeyDown).toHaveBeenCalledWith(expect.any(Object))
  expect(onClick).toHaveBeenCalledWith(expect.any(Object))
  expect(onScroll).not.toHaveBeenCalled()
})

test('calls onKeyDown and onClick when keyCode is 13', () => {
  const onClick = jest.fn()
  const onKeyDown = jest.fn()
  const { getByText } = render(
    <ButtonA11y
      className="mock-className"
      onKeyDown={onKeyDown}
      onClick={onClick}
      type=""
    >
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { keyCode: 13 })

  expect(onKeyDown).toHaveBeenCalledWith(expect.any(Object))
  expect(onClick).toHaveBeenCalledWith(expect.any(Object))
})

test('calls onKeyDown and onClick when keyCode is 32', () => {
  const onClick = jest.fn()
  const onKeyDown = jest.fn()
  const { getByText } = render(
    <ButtonA11y
      className="mock-className"
      onKeyDown={onKeyDown}
      onClick={onClick}
      type=""
    >
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { keyCode: 32 })

  expect(onKeyDown).toHaveBeenCalledWith(expect.any(Object))
  expect(onClick).toHaveBeenCalledWith(expect.any(Object))
})

test('only calls onKeyDown when random key', () => {
  const onClick = jest.fn()
  const onKeyDown = jest.fn()
  const { getByText } = render(
    <ButtonA11y
      className="mock-className"
      onKeyDown={onKeyDown}
      onClick={onClick}
      type=""
    >
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { key: 'Delete' })

  expect(onKeyDown).toHaveBeenCalledWith(expect.any(Object))
  expect(onClick).not.toHaveBeenCalled()
})

test('only calls onKeyDown when random keyCode', () => {
  const onClick = jest.fn()
  const onKeyDown = jest.fn()
  const { getByText } = render(
    <ButtonA11y
      className="mock-className"
      onKeyDown={onKeyDown}
      onClick={onClick}
      type=""
    >
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { keyCode: 17 })

  expect(onKeyDown).toHaveBeenCalledWith(expect.any(Object))
  expect(onClick).not.toHaveBeenCalled()
})

test('does not break when onKeyDown not passed', () => {
  const onClick = jest.fn()
  const { getByText } = render(
    <ButtonA11y className="mock-className" onClick={onClick} type="">
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { key: 'Enter' })

  expect(onClick).toHaveBeenCalledWith(expect.any(Object))
})

test('does not break when onClick not passed', () => {
  const onKeyDown = jest.fn()
  const { getByText } = render(
    <ButtonA11y className="mock-className" onKeyDown={onKeyDown} type="">
      Click Me
    </ButtonA11y>
  )
  const el = getByText('Click Me')

  expect(el).toBeInTheDocument()

  fireEvent.keyDown(el, { key: 'Enter' })

  expect(onKeyDown).toHaveBeenCalledWith(expect.any(Object))
})

test('renders as div element', () => {
  const { container, getByText } = render(
    <ButtonA11y className="mock-className" element="div" onClick={jest.fn()}>
      Click Me
    </ButtonA11y>
  )

  expect(getByText('Click Me')).toBeInTheDocument()
  expect(container.firstChild.tagName).toEqual('DIV')
})

test('forwards a ref', () => {
  const ref = createRef()
  render(
    <ButtonA11y className="mock-className" onClick={jest.fn()} ref={ref}>
      Click Me
    </ButtonA11y>
  )

  expect(ref.current).toBeInTheDocument()
})

test('errors when no children & no aria-label', () => {
  expect(() => {
    render(
      <ButtonA11y
        className="mock-className"
        element="span"
        onClick={jest.fn()}
      />
    )
  }).toThrow('react-button-a11y: `aria-label` required for accessibility')
})

// This is allowed but is unwise for accessibility
test('renders when not strictMode, no children & no aria-label', () => {
  const { getByRole } = render(
    <ButtonA11y
      className="mock-className"
      element="span"
      onClick={jest.fn()}
      strictMode={false}
    />
  )
  const el = getByRole('button')

  expect(el).toBeInTheDocument()
})


test/index.test.js
import React, { createRef } from 'react'
import { renderIntoDocument } from 'react-dom/test-utils'
import { mount, shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Container, { ButtonA11y } from '../source'

describe('ButtonA11y', () => {

  describe('container', () => {

    it('renders', () => {
      const wrapper = shallow(
        <Container
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('handles ref', () => {
      const ref = createRef()
      renderIntoDocument(
        <Container
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
          ref={ref}
        />
      )
      expect(ref.current).toBeInstanceOf(HTMLSpanElement)
    })

  })

  describe('component', () => {

    it('renders, no children', () => {
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('renders, children', () => {
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
        >
          Click me
        </ButtonA11y>
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('renders, span element', () => {
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          element="span"
          onClick={jest.fn()}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('renders, strict mode disabled', () => {
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
          strictMode={false}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('renders, ref passed in', () => {
      const ref = createRef()
      const wrapper = mount(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          forwardedRef={ref}
          onClick={jest.fn()}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('errors when no children AND aria-label not passed', () => {
      const render = () =>
        shallow(
          <ButtonA11y
            className="mock-className"
            element="span"
            onClick={jest.fn()}
          />
        )

      expect(render).toThrow(
        'react-button-a11y: `aria-label` required for accessibility'
      )
    })

    it('does not break when onClick not passed', () => {
      const onKeyDown = jest.fn()
      const preventDefault = jest.fn()
      const e = { key: 'Enter', preventDefault }
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          element="span"
          onKeyDown={onKeyDown}
        />
      )

      wrapper.first().prop('onKeyDown')(e)

      expect(preventDefault).toHaveBeenCalled()
      expect(onKeyDown).toHaveBeenCalledWith(e)
    })

    it('does not break when onKeyDown not passed', () => {
      const onClick = jest.fn()
      const preventDefault = jest.fn()
      const e = { key: 'Enter', preventDefault }
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          element="span"
          onClick={onClick}
        />
      )

      wrapper.first().prop('onKeyDown')(e)

      expect(preventDefault).toHaveBeenCalled()
      expect(onClick).toHaveBeenCalledWith(e)
    })

    it('does not break when onKeyDown AND onClick not passed', () => {
      const preventDefault = jest.fn()
      const e = { key: 'Enter', preventDefault }
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          element="span"
        />
      )

      wrapper.first().prop('onKeyDown')(e)

      expect(preventDefault).toHaveBeenCalled()
    })

    describe('when random keydown key', () => {

      it('only calls onKeyDown', () => {
        const onClick = jest.fn()
        const onKeyDown = jest.fn()
        const preventDefault = jest.fn()
        const e = { key: 'ArrowDown', preventDefault }
        const wrapper = shallow(
          <ButtonA11y
            aria-label="Mock Label"
            className="mock-className"
            onClick={onClick}
            onKeyDown={onKeyDown}
          />
        )

        wrapper.first().prop('onKeyDown')(e)

        expect(preventDefault).not.toHaveBeenCalled()
        expect(onClick).not.toHaveBeenCalled()
        expect(onKeyDown).toHaveBeenCalledWith(e)
      })

    })

    describe('when random keydown keyCode', () => {

      it('only calls onKeyDown', () => {
        const onClick = jest.fn()
        const onKeyDown = jest.fn()
        const preventDefault = jest.fn()
        const e = { keyCode: 10, preventDefault }
        const wrapper = shallow(
          <ButtonA11y
            aria-label="Mock Label"
            className="mock-className"
            onClick={onClick}
            onKeyDown={onKeyDown}
          />
        )

        wrapper.first().prop('onKeyDown')(e)

        expect(preventDefault).not.toHaveBeenCalled()
        expect(onClick).not.toHaveBeenCalled()
        expect(onKeyDown).toHaveBeenCalledWith(e)
      })

    })

    describe('when keydown is enter-like', () => {

      describe('when key of Enter', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { key: 'Enter', preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

      describe('when keyCode of 13', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { keyCode: 13, preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

    })

    describe('when keydown is spacebar-like', () => {

      describe('when key of " "', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { key: ' ', preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

      describe('when key of Spacebar', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { key: 'Spacebar', preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

      describe('when keyCode of 32', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { keyCode: 32, preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

    })

  })

})


test/react-button-a11y.test.js
import React, { createRef } from 'react'
import { renderIntoDocument } from 'react-dom/test-utils'
import { mount, shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Container, { ButtonA11y } from '../source/react-button-a11y'

describe('react-button-a11y', () => {

  describe('container', () => {

    it('renders', () => {
      const wrapper = shallow(
        <Container
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('handles ref', () => {
      const ref = createRef()
      renderIntoDocument(
        <Container
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
          ref={ref}
        />
      )
      expect(ref.current).toBeInstanceOf(HTMLDivElement)
    })

  })

  describe('component', () => {

    it('renders, no children', () => {
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('renders, children', () => {
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
        >
          Click me
        </ButtonA11y>
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('renders, span element', () => {
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          element="span"
          onClick={jest.fn()}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('renders, strict mode disabled', () => {
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          onClick={jest.fn()}
          strictMode={false}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('renders, ref passed in', () => {
      const ref = createRef()
      const wrapper = mount(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          forwardedRef={ref}
          onClick={jest.fn()}
        />
      )
      const tree = toJson(wrapper)
      expect(tree).toMatchSnapshot()
    })

    it('errors when no children AND aria-label not passed', () => {
      const render = () =>
        shallow(
          <ButtonA11y
            className="mock-className"
            element="span"
            onClick={jest.fn()}
          />
        )

      expect(render).toThrow(
        'react-button-a11y: `aria-label` required for accessibility'
      )
    })

    it('does not break when onClick not passed', () => {
      const onKeyDown = jest.fn()
      const preventDefault = jest.fn()
      const e = { key: 'Enter', preventDefault }
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          element="span"
          onKeyDown={onKeyDown}
        />
      )

      wrapper.first().prop('onKeyDown')(e)

      expect(preventDefault).toHaveBeenCalled()
      expect(onKeyDown).toHaveBeenCalledWith(e)
    })

    it('does not break when onKeyDown not passed', () => {
      const onClick = jest.fn()
      const preventDefault = jest.fn()
      const e = { key: 'Enter', preventDefault }
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          element="span"
          onClick={onClick}
        />
      )

      wrapper.first().prop('onKeyDown')(e)

      expect(preventDefault).toHaveBeenCalled()
      expect(onClick).toHaveBeenCalledWith(e)
    })

    it('does not break when onKeyDown AND onClick not passed', () => {
      const preventDefault = jest.fn()
      const e = { key: 'Enter', preventDefault }
      const wrapper = shallow(
        <ButtonA11y
          aria-label="Mock Label"
          className="mock-className"
          element="span"
        />
      )

      wrapper.first().prop('onKeyDown')(e)

      expect(preventDefault).toHaveBeenCalled()
    })

    describe('when random keydown key', () => {

      it('only calls onKeyDown', () => {
        const onClick = jest.fn()
        const onKeyDown = jest.fn()
        const preventDefault = jest.fn()
        const e = { key: 'ArrowDown', preventDefault }
        const wrapper = shallow(
          <ButtonA11y
            aria-label="Mock Label"
            className="mock-className"
            onClick={onClick}
            onKeyDown={onKeyDown}
          />
        )

        wrapper.first().prop('onKeyDown')(e)

        expect(preventDefault).not.toHaveBeenCalled()
        expect(onClick).not.toHaveBeenCalled()
        expect(onKeyDown).toHaveBeenCalledWith(e)
      })

    })

    describe('when random keydown keyCode', () => {

      it('only calls onKeyDown', () => {
        const onClick = jest.fn()
        const onKeyDown = jest.fn()
        const preventDefault = jest.fn()
        const e = { keyCode: 10, preventDefault }
        const wrapper = shallow(
          <ButtonA11y
            aria-label="Mock Label"
            className="mock-className"
            onClick={onClick}
            onKeyDown={onKeyDown}
          />
        )

        wrapper.first().prop('onKeyDown')(e)

        expect(preventDefault).not.toHaveBeenCalled()
        expect(onClick).not.toHaveBeenCalled()
        expect(onKeyDown).toHaveBeenCalledWith(e)
      })

    })

    describe('when keydown is enter-like', () => {

      describe('when key of Enter', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { key: 'Enter', preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

      describe('when keyCode of 13', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { keyCode: 13, preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

    })

    describe('when keydown is spacebar-like', () => {

      describe('when key of " "', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { key: ' ', preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

      describe('when key of Spacebar', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { key: 'Spacebar', preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

      describe('when keyCode of 32', () => {

        it('prevents default & calls onClick callback', () => {
          const onClick = jest.fn()
          const onKeyDown = jest.fn()
          const preventDefault = jest.fn()
          const e = { keyCode: 32, preventDefault }
          const wrapper = shallow(
            <ButtonA11y
              aria-label="Mock Label"
              className="mock-className"
              onClick={onClick}
              onKeyDown={onKeyDown}
            />
          )

          wrapper.first().prop('onKeyDown')(e)

          expect(preventDefault).toHaveBeenCalled()
          expect(onClick).toHaveBeenCalledWith(e)
          expect(onKeyDown).toHaveBeenCalledWith(e)
        })

      })

    })

  })

})


example/app.js
import React, { useCallback, useState } from 'react'
import ReactDOM from 'react-dom'
import ButtonA11y from '../dist/cjs'

const App = () => {
  const [isBtn1Pressed, pressBtn1] = useState(false)
  const [isBtn2Pressed, pressBtn2] = useState(false)

  const handleBtn1Click = useCallback(() => pressBtn1(x => !x), [])
  const handleBtn2Click = useCallback(() => pressBtn2(x => !x), [])

  const btn1Label = isBtn1Pressed ? 'Pressed' : 'Press me'
  const btn1Class = 'btn--words'.concat(isBtn1Pressed ? ' btn--pressed' : '')

  const btn2Label = isBtn2Pressed ? 'Pressed' : 'Press me'
  const btn2Class = 'btn--noWords'.concat(isBtn2Pressed ? ' btn--pressed' : '')

  return (
    <div>
      <section>
        <h2>Button with children</h2>
        <ButtonA11y className={btn1Class} onClick={handleBtn1Click}>
          {btn1Label}
        </ButtonA11y>
      </section>
      <section>
        <h2>Button without children</h2>
        <ButtonA11y
          aria-label={btn2Label}
          className={btn2Class}
          onClick={handleBtn2Click}
        />
      </section>
    </div>
  )
}

const container = document.querySelector('[data-app]')
ReactDOM.render(<App />, container)


source/index.js
import { createElement, memo, useCallback } from 'react'
import withForwardedRef from 'react-with-forwarded-ref'

// isSelect :: KeyboardEvent -> Bool
const isSelect = e =>
  e.key === 'Enter' ||
  e.key === ' ' ||
  e.key === 'Spacebar' ||
  e.keyCode === 13 ||
  e.keyCode === 32

const ButtonA11y = ({
  children,
  element,
  forwardedRef,
  onClick,
  onKeyDown,
  strictMode,
  ...rest
}) => {
  if (strictMode && !children && !rest['aria-label']) {
    throw new Error(
      'react-button-a11y: `aria-label` required for accessibility'
    )
  }

  const handleKeyDown = useCallback(
    e => {
      onKeyDown(e)

      if (isSelect(e)) {
        e.preventDefault()
        onClick(e)
      }
    },
    [onClick, onKeyDown]
  )

  return createElement(element, {
    ...rest,
    children,
    onClick,
    onKeyDown: handleKeyDown,
    ref: forwardedRef,
    role: 'button',
    tabIndex: '0'
  })
}

ButtonA11y.defaultProps = {
  element: 'span',
  onClick: Function.prototype,
  onKeyDown: Function.prototype,
  strictMode: true
}

export default memo(withForwardedRef(ButtonA11y))
